### PR TITLE
Add incoming and outgoing packet types to chain context

### PR DESCRIPTION
**Pull request by soareschen**
Closes: #2840

Blocked by: #2957

## Description

- Introduce a new `HasIbcPacketTypes` trait that contains `IncomingPacket` and `OutgoingPacket` types.
- Reimplement `CanBuildReceivePacketMessage` using only chain context without requiring relay context.
- The remaining relay context methods can be refactored in similar way in subsequent PRs.

@ljoss17 this is currently blocked by #2957, as the chain context do not have information about the client ID when building the receive packet message. Feel free to push changes directly to this branch.

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/relayer-framework/src/base/chain/traits/types/ibc.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+use crate::base::core::traits::sync::Async;
+use crate::std_prelude::*;
+
+/**
+   The abstract types for a chain context when it is used for IBC
+   communication with a `Counterparty` chain context.
+
+   In contrast to [`HasChainTypes`], this trait is parameterized by a
+   `Counterparty` chain context, which is also required to implement
+   [`HasChainTypes`].
```
Comment (seanchen1991) [inline crates/relayer-framework/src/base/chain/traits/types/ibc.rs:11]: Not sure if there's a typo here or not, but this is confusing: I read it as saying that `HasChainTypes` is _not_ parameterized by a `Counterparty` chain context, but that `HasChainTypes` requires a `Counterparty` chain context in order to be implemented.

**Review by seanchen1991 — COMMENTED** [review]
I see that this PR is blocked, but it provides a lot of great documentation. It would great if as much of the documentation could be merged in, perhaps in a separate PR?

File: crates/relayer-framework/src/base/chain/traits/types/ibc.rs (hunk 1 above)
Comment (soareschen) [reply crates/relayer-framework/src/base/chain/traits/types/ibc.rs]: I rephrased it to two separate sentences to make it clear that it is `Counterparty` that is required to implement `HasChainTypes`.

**Comment by soareschen** [comment]
Actually the documentation are from the base branch, and I just split the constructs up into separate files.

But you are right that we can merge this PR first and fix the mock chain in separate PR. I already disabled the test so the CI is passing for now. Merging first can also reduce the chance of conflict in case anyone of us changed the code or documentation.

**Pull request merged**